### PR TITLE
Ch06 建立群組系統

### PR DESCRIPTION
**Pull request by Fatmylin**
### Feature 
- Add: group relation between groups and users
- Add: functions: `member_of? `, `join!` ,` quit!` and `admin_of?`
- Add: before_action `should_be_post_owner`

### How to use
-  run rails s
- Open localhost:3000
- try to create a discuss group
- try to join a group
- try to quit a group

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -42,6 +42,7 @@ gem 'erb_lint', '~> 0.0.26'
 gem 'rubocop', '~> 0.51.0'
 # will_paginate
 gem 'will_paginate', '~> 3.1.6'
+gem 'awesome_print'
```
Comment (kakas) [inline Gemfile]: 這個應該是開發用的套件吧？
可以放到 `group :development` 下就好了

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -44,4 +47,8 @@ def should_have_post
 
     head(404) if @post.nil?
   end
+
+  def should_be_post_owner
+    redirect_to group_path(@group) if current_user.id != @post.user.id
```
Comment (kakas) [inline app/controllers/posts_controller.rb]: `@post.user_id`

File: app/models/group.rb [hunk 3]
```diff
@@ -3,4 +3,6 @@ class Group < ApplicationRecord
 
   belongs_to :user
   has_many :posts, dependent: :destroy
+  has_many :group_relationships
```
Comment (kakas) [inline app/models/group.rb]: 覺得叫做 `:group_users` 比較明確

File: app/models/user.rb [hunk 4]
```diff
@@ -7,4 +7,22 @@ class User < ApplicationRecord
...
+  def join!(group)
+    participated_groups << group
+  end
+
+  def quit!(group)
+    participated_groups.delete(group)
+  end
+
+  def admin_of?(group)
+    id == group.user.id
```
Comment (kakas) [inline app/models/user.rb]: `user_id`

**Review by DaMinaup6 — APPROVED** [review]

**Review by glasses618 — APPROVED** [review]

**Pull request merged**